### PR TITLE
Drop support for kubernetes < 1.20 for GCP extension provider

### PR DESCRIPTION
**Pull request by dimitar-kostadinov**
**How to categorize this PR?**

/area open-source
/kind cleanup
/platform gcp

**What this PR does / why we need it**:
Drop support for kubernetes < 1.20 for GCP extension provider

**Which issue(s) this PR fixes**:
Part of https://github.com/gardener/gardener/issues/6911

**Special notes for your reviewer**:

**Release note**:

```breaking operator
provider-gcp no longer supports Shoots with Кubernetes version < 1.20.
```

### CODE REVIEW

File: pkg/webhook/controlplane/ensurer.go [hunk 1]
```diff
@@ -188,71 +163,40 @@ func (e *ensurer) EnsureClusterAutoscalerDeployment(ctx context.Context, gctx gc
...
+	c.Command = extensionswebhook.EnsureStringWithPrefixContains(c.Command, "--feature-gates=",
+		"CSIMigrationGCE=true", ",")
+	c.Command = extensionswebhook.EnsureStringWithPrefixContains(c.Command, "--feature-gates=",
+		csiMigrationCompleteFeatureGate+"=true", ",")
+	c.Command = extensionswebhook.EnsureNoStringWithPrefix(c.Command, "--cloud-provider=")
+	c.Command = extensionswebhook.EnsureNoStringWithPrefix(c.Command, "--cloud-config=")
+	c.Command = extensionswebhook.EnsureNoStringWithPrefixContains(c.Command, "--enable-admission-plugins=",
 		"PersistentVolumeLabel", ",")
-	c.Command = extensionswebhook.EnsureNoStringWithPrefixContains(c.Command, "--disable-admission-plugins=",
+	c.Command = extensionswebhook.EnsureStringWithPrefixContains(c.Command, "--disable-admission-plugins=",
```
Comment (MartinWeindel) [inline pkg/webhook/controlplane/ensurer.go:177]: Strange diff, but correct (see original line 205) :-)

**Review by MartinWeindel — CHANGES_REQUESTED** [review]
Thanks for the nice PR. I already made use of it as a template for preparing a similar PR for provider-aws.

I have only one minor change request.

**Review by MartinWeindel — APPROVED** [review]
/lgtm

**Review by MartinWeindel — APPROVED** [review]
/lgtm

File: pkg/webhook/controlplane/ensurer.go [hunk 2]
```diff
@@ -50,13 +47,11 @@ func NewEnsurer(logger logr.Logger) genericmutator.Ensurer {
...
 type ensurer struct {
 	genericmutator.NoopEnsurer
-	client client.Client
 	logger logr.Logger
 }
 
 // InjectClient injects the given client into the ensurer.
-func (e *ensurer) InjectClient(client client.Client) error {
-	e.client = client
+func (e *ensurer) InjectClient(_ client.Client) error {
```
Comment (ialidzhikov) [inline pkg/webhook/controlplane/ensurer.go:54]: You could remove the whole `InjectClient` func as the ensurer no longer needs a client.

**Review by ialidzhikov — CHANGES_REQUESTED** [review]

File: pkg/webhook/controlplane/ensurer.go [hunk 3]
```diff
@@ -357,92 +284,42 @@ var (
...
-		c.VolumeMounts = extensionswebhook.EnsureNoVolumeMountWithName(c.VolumeMounts, cloudProviderConfigVolumeMount.Name)
-		c.VolumeMounts = extensionswebhook.EnsureNoVolumeMountWithName(c.VolumeMounts, cloudProviderSecretVolumeMount.Name)
-		return
-	}
-
-	c.VolumeMounts = extensionswebhook.EnsureVolumeMountWithName(c.VolumeMounts, cloudProviderConfigVolumeMount)
-	c.VolumeMounts = extensionswebhook.EnsureVolumeMountWithName(c.VolumeMounts, cloudProviderSecretVolumeMount)
+func ensureKubeAPIServerVolumeMounts(c *corev1.Container) {
+	c.VolumeMounts = extensionswebhook.EnsureNoVolumeMountWithName(c.VolumeMounts, cloudProviderConfigVolumeMount.Name)
+	c.VolumeMounts = extensionswebhook.EnsureNoVolumeMountWithName(c.VolumeMounts, cloudProviderSecretVolumeMount.Name)
```
Comment (ialidzhikov) [inline pkg/webhook/controlplane/ensurer.go:289]: Assuming that all clusters are now running at least K8s 1.20 and the CSI migration was complete for them in 1.18 we can clean up these `EnsureNoVolumeMountWithName` calls.

Previously the handling was:
https://github.com/gardener/gardener-extension-provider-gcp/blob/38db4bac5eadeba6b5aa7f4763e09fe87cde4416/pkg/webhook/controlplane/ensurer.go#L360-L369

Prior to CSI we were ensuring volume and volume mount for the cloud-provider-config ConfigMap for components like the kube-apiserver, kube-controller-manager. With CSI we no longer need this volume and volume mount, hence we were ensuring that any potential present volume/volume mount gets cleaned up after the update to a version that enables CSI.
I think it should be safe to drop these `EnsureNoVolumeMountWithName` calls for the cloud-provider-config ConfigMap as they should be already cleaned up from the kube-apiserver/kube-controller-manager deployments after CSI migration is complete. And today there shouldn't be any handling that adds these cloud-provider-config volume and volume mount.

File: pkg/webhook/controlplane/ensurer.go [hunk 4]
```diff
@@ -357,92 +284,42 @@ var (
...
-
-	c.VolumeMounts = extensionswebhook.EnsureVolumeMountWithName(c.VolumeMounts, cloudProviderConfigVolumeMount)
-	c.VolumeMounts = extensionswebhook.EnsureVolumeMountWithName(c.VolumeMounts, cloudProviderSecretVolumeMount)
-
-	c.VolumeMounts = extensionswebhook.EnsureVolumeMountWithName(c.VolumeMounts, etcSSLVolumeMount)
-	// some distros have symlinks from /etc/ssl/certs to /usr/share/ca-certificates
-	c.VolumeMounts = extensionswebhook.EnsureVolumeMountWithName(c.VolumeMounts, usrShareCaCertsVolumeMount)
+func ensureKubeControllerManagerVolumeMounts(c *corev1.Container) {
+	c.VolumeMounts = extensionswebhook.EnsureNoVolumeMountWithName(c.VolumeMounts, cloudProviderConfigVolumeMount.Name)
+	c.VolumeMounts = extensionswebhook.EnsureNoVolumeMountWithName(c.VolumeMounts, cloudProviderSecretVolumeMount.Name)
```
Comment (ialidzhikov) [inline pkg/webhook/controlplane/ensurer.go:294]: Same as the above comment for the kube-apiserver.

File: pkg/webhook/controlplane/ensurer.go [hunk 5]
```diff
@@ -357,92 +284,42 @@ var (
...
-	if csiEnabled && csiMigrationComplete {
-		if template.Annotations != nil {
-			delete(template.Annotations, "checksum/secret-"+v1beta1constants.SecretNameCloudProvider)
-			delete(template.Annotations, "checksum/configmap-"+internal.CloudProviderConfigName)
-		}
-		return nil
+func (e *ensurer) ensureChecksumAnnotationsDeleted(template *corev1.PodTemplateSpec) {
+	if template.Annotations != nil {
+		delete(template.Annotations, "checksum/secret-"+v1beta1constants.SecretNameCloudProvider)
+		delete(template.Annotations, "checksum/configmap-"+internal.CloudProviderConfigName)
```
Comment (ialidzhikov) [inline pkg/webhook/controlplane/ensurer.go:314]: This handling also could be removed when we drop the `EnsureNoVolumeMountWithName` handling for the cloud-provider-config ConfigMap.

**Review by ialidzhikov — COMMENTED** [review]
Looks good! Thank you for the clean up! Minor suggestion inline.

File: pkg/controller/controlplane/valuesprovider.go [hunk 6]
```diff
@@ -328,40 +327,19 @@ func (vp *valuesProvider) GetControlPlaneShootChartValues(
...
 ) (map[string]interface{}, error) {
-	k8sVersionLessThan118, err := version.CompareVersions(cluster.Shoot.Spec.Kubernetes.Version, "<", gcp.CSIMigrationKubernetesVersion)
-	if err != nil {
-		return nil, err
-	}
-
 	return map[string]interface{}{
 		"volumesnapshots": map[string]interface{}{
-			"enabled": !k8sVersionLessThan118,
+			"enabled": true,
```
Comment (ialidzhikov) [inline pkg/controller/controlplane/valuesprovider.go:338]: Is it possible to change the default value to true here
https://github.com/gardener/gardener-extension-provider-gcp/blob/310dd84ac7159cb4009d429bf55d9af5e48ddb86/charts/internal/shoot-crds/values.yaml#L1-L2
and to drop this `GetControlPlaneShootCRDsChartValues` func?

We could also adapt https://github.com/gardener/gardener-extension-provider-gcp/blob/9c55612de3c5bf27e4f7086a4b39d826aa7b7e30/charts/internal/seed-controlplane/values.yaml#L6-L7 - the csi-driver-controller chart to be enabled by default.

**Review by ialidzhikov — APPROVED** [review]
/lgtm

**Review by kon-angelo — APPROVED** [review]
/lgtm
/reviewed ok-to-test

**Review by kon-angelo — APPROVED** [review]
/lgtm

**Review by ialidzhikov — APPROVED** [review]
/lgtm

**Pull request merged**